### ISSUE TITLE
Performance regression

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#4771]
In PR #2585 I attempted to rewrite `EvalString` in a "cleaner" fashion. Back then this was rejected because it was reported to slow down things by a factor of three. Today I discussed this approach with @laurentbartholdi (he suggested the same refactoring of the code independently), which lead me to reexamine the performance issue... And I think we have a performance regression here.

So here is my test code, comparing "old" and "new" `EvalString` implementation:
```gap
myeval := function( s )
  local a, f;
  # Prepend 'return' and append ';' to the given string; also add newlines to
  # ensure that if there are syntax errors in the evaluated string, GAP
  # prints only the evaluated string, but not the 'return' and ';' we prepend
  # resp. append.
  a := "return\n";
  Append(a, s);
  Append(a, ";\n");
  f := ReadAsFunction(InputTextString(a));
  return f();
end;;
for i in [1..100000] do EvalString("12+13"); od; time;
for i in [1..100000] do myeval("12+13"); od; time;
```

Some timings done on my MacBok Pro 14" with M1 CPU:
```
# GAP master
gap> for i in [1..100000] do EvalString("12+13"); od; time;
572
gap> for i in [1..100000] do myeval("12+13"); od; time;
552

# GAP 4.11.1
gap> for i in [1..100000] do EvalString("12+13"); od; time;
360
gap> for i in [1..100000] do myeval("12+13"); od; time;
338

# GAP 4.10.2
gap> for i in [1..100000] do EvalString("12+13"); od; time;
128
gap> for i in [1..100000] do myeval("12+13"); od; time;
448

# GAP 4.9.3
gap> for i in [1..100000] do EvalString("12+13"); od; time;
125
gap> for i in [1..100000] do myeval("12+13"); od; time;
327

# GAP 4.8
for i in [1..100000] do EvalString("12+13"); od; time;
83
for i in [1..100000] do myeval("12+13"); od; time;
96
```
So when looking at GAP 4.9 or 4.10, one really sees a factor of 2.5 - 3.5. But in GAP master and in GAP 4.8 both methods are about the same speed -- except GAP 4.8 is faster by a factor of ~5.5...

After some more experimenting, I noticed the following: if I run the second loop multiple times in GAP 4.8, it gets slower, until it stabilizes at about ~350 (which is about the same as in GAP 4.11); even if I force a garbage collection. This is really strange:
```
gap> GASMAN("collect"); for i in [1..100000] do myeval("12+13"); od; time;
324
gap> GASMAN("collect"); for i in [1..100000] do myeval("12+13"); od; time;
323
gap> GASMAN("collect"); for i in [1..100000] do myeval("12+13"); od; time;
377
gap> GASMAN("collect"); for i in [1..100000] do myeval("12+13"); od; time;
343
gap> GASMAN("collect"); for i in [1..100000] do myeval("12+13"); od; time;
347
```
No such slowdown happened for `EvalString`.

**Comment by fingolfin** [comment]
Just a quick update: the problem is not related to the constructing of the string or the stream (not that I thought it was, but it's good to exclude it). I tested via this code:
```gap
stream:=InputTextString("return 12+13;\n");
myeval3 := function()
  local f;
  RewindStream(stream);
  f := ReadAsFunction(stream);
  return f();
end;;
for i in [1..100000] do myeval3(); od; time;
```
And got this:
```
# GAP master
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
378
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
376
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
377
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
376
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
377

# GAP 4.8
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
83
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
90
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
98
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
107
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
122
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
147
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
188
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
188
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
327
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
204
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
452
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
221
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
224
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
824
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
248
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
260
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
1370
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
273
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
283
```

**Comment by fingolfin** [comment]
Yet with GAP 4.7 the weird slowdown does not happen... hurgh:
```
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
80
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
78
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
80
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
79
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
79
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
80
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
79
gap> GASMAN("collect"); for i in [1..100000] do myeval3(); od; time;
78
```

**Comment by ChrisJefferson** [comment]
Looking at the profile, nothing massive jumps out, but a bunch of small things do. I think we do a bit more work now when we set up a "SHELL" for parsing (for example making a TypInputFile requires setting the whole thing to 0, which includes a 32K buffer for 'line' -- I'm not sure I totally true my profile as some things are a bit weird, but it suggests that is taking about 30% of the time). There is also a bunch of code related to profiling to set line+filename for parsed tokens, which is taking another 20%

We made sure this stuff didn't slow down things like loading GAP and it's packages, but here basically all the time is being spent setting up, doing, and tearing down parsing.

**Comment by ChrisJefferson** [comment]
The other expensive part is GetInputFilenameID (io.c:257-278), which uses linear search to look up the filename in a list of all previously parsed files.

**Comment by fingolfin** [comment]
Thanks for tracking these down, I'll work on fixing this

**Comment by fingolfin** [comment]
I've addressed the first in PR #4772. For `GetInputFilenameID` I can confirm that disabling it restores us to about original performance. I've started working on a "proper" replacement, but it requires implementing yet another hashtable, so instead I am now factoring out the common hashtable code in gvars.c
